### ISSUE TITLE
Packaging: contribution well welcomed

### ISSUE THREAD

**Issue by phaazon** [phaazon/kak-tree-sitter#22]
Currently, there is no release channel for people to install the project, besides `crates.io`. It’s fine for people working with Rust, but I stand by the idea that binaries on `crates.io` should only be about Rust development, and hence, I don’t think it’s a good release channel.

`kak-lsp` is in the `extra` Archlinux repository, for instance, and I think it would be great to have a similar thing for `kak-tree-sitter`. Same for `brew`.

I haven’t really thought about how to ease that process to package managers, and it’s a bit early, but it’s never too early to start thinking about that.

On the meantime, I’ll bootstrap a `PKGBUILD` for the AUR.

**Comment by rosingrind** [comment]
I'm using macOS and `brew` package will be a nice addition btw. If I understand correctly, `brew` is all about `formula`s or `cask`s (terminal package or GUI macOS `.app`), you'll need to host a `formula`.

Formulas are hosted either in main `brew` registry, or in a `tap` repo (prefixed with `homebrew-*`), `tap`s may either be added manually by user with `brew tap` or looked up by brew during install. I think you'd like to use this as a reference for creating your `tap`: https://github.com/kak-lsp/homebrew-kak-lsp

**Comment by rosingrind** [comment]
By the way, I'm using `kak-bundle` to manage my dependencies and plugins (same as `plug` but faster and maintained), and I'd like to manage `kak-tree-sitter` with it (using git tags), is it possible? As far as I know, I will need to run 2 post-install commands:

- `cargo install --locked --force --path ./kak-tree-sitter`
- `cargo install --locked --force --path ./ktsctl`

Is this appropriate or it's better to just install it from cargo registry?

**Comment by phaazon** [comment]
So thar it’s good, but I plan on doing github releases as well as git tags. Do need something specific?

**Comment by rosingrind** [comment]
Everything is ok, no specific moves needed, but I think it may be better to split packages to separate repos in future as they'll definitely grow. If you don't mind, I'll try to take on creating brew packages for you!

**Comment by rosingrind** [comment]
> I'll try to take on creating brew packages

Added you as a collaborator on https://github.com/rosingrind/homebrew-kak-tree-sitter, basically it's 2 formulas with prebuilt x86_64 bottles (binary packages) via github actions. I'll transfer these to you if you'd like to - there is an unsolved issue with bottling an apple silicon binaries, which may require transferring these formulas to `homebrew-core`, and I think it's up to you to make a desicion there

**Comment by phaazon** [comment]
All good, thanks a lot! I’ll add that to the wiki!

**Comment by jdugan6240** [comment]
> By the way, I'm using kak-bundle to manage my dependencies and plugins (same as plug but faster and maintained), and I'd like to manage kak-tree-sitter with it (using git tags), is it possible? As far as I know, I will need to run 2 post-install commands:

Sorry if I'm seeing this too late, but that should work, I think.

**Comment by rosingrind** [comment]
@jdugan6240 yeah, I'm using something like this and it definitely works well

```kakounescript
bundle-customload kak-tree-sitter 'git clone -b kak-tree-sitter-v0.4.4 https://github.com/phaazon/kak-tree-sitter.git' %{
	eval %sh{ kak-tree-sitter -dsk --session $kak_session }
	colorscheme pastel
	...
} %{
	cargo install --locked --force --path ./kak-tree-sitter
}
```

**Comment by uncomfyhalomacro** [comment]
i want to package this to opensuse, but I am confused on how to package this, @phaazon are `ktsctl` and `kak-tree-sitter` separate packages *and* have separate versions?

**Comment by jdugan6240** [comment]
> i want to package this to opensuse, but I am confused on how to package this, @phaazon are `ktsctl` and `kak-tree-sitter` separate packages *and* have separate versions?

Yes, they are. `kak-tree-sitter` is currently 0.5.0-dev (basically a snapshot), and `ktsctl` is 0.3.3-dev. Thus, they do need to be packaged separately.

**Comment by phaazon** [comment]
Yes they are two different tools. KTS is the server and `ktsctl` is just a controller binary to help organizing resources, and it’s optional.

**Comment by igor-ramazanov** [comment]
There's a [PR](https://github.com/phaazon/kak-tree-sitter/pull/200) to package as a nix flake together with grammars and queries from Helix.

**Comment by caksoylar** [comment]
Are there still plans to create binary Github releases? I'd appreciate it, as I don't have a Rust toolchain on all my computers and kakoune-lsp publishing releases with binaries is very helpful for me.

**Comment by krobelus** [comment]
Sorry if this is out of scope here - FWIW I think it would make sense to rename the packages (not the binaries) to something like `kakoune-tree-sitter{,-ctl}`. This matches Kakoune; also kak without context is a bad word in many languages.

I think most distributions have a concept of "package A replaces package B".
I'm not sure how it works with crates.io, I don't think we can delete a package but maybe yank all versions and have the readme point to a new one?
I'm guessing that means users who use the old name get a warning, never an error?

**Comment by phaazon** [comment]
> Are there still plans to create binary Github releases? I'd appreciate it, as I don't have a Rust toolchain on all my computers and kakoune-lsp publishing releases with binaries is very helpful for me.

Yes, it’s planned. At least for `v1.0`; before if I can make it. I need to think a bit more about the process. I need to make a CD workflow to create the releases and sign them, which is not something I have thought about. I want to make things right, so it would require a proper signature of the binaries, at least, so that people can properly ensure they have downloaded the right thing.

---

> Sorry if this is out of scope here - FWIW I think it would make sense to rename the packages (not the binaries) to something like kakoune-tree-sitter{,-ctl}. This matches Kakoune; also kak without context is a bad word in many languages.

We discussed that on Discord and indeed, maybe it could be a good idea to move, if that’s what the community wants. I don’t think I will rename `ktsctl`, though. The name is perfectly fine to me.

> I'm not sure how it works with crates.io, I don't think we can delete a package but maybe yank all versions and have the readme point to a new one?

So, yanking doesn’t remove; it just invalidates, so if someone still has a build plan (i.e. `Cargo.lock`), they will still be able to download it. However, if you start a new project with it, it won’t be found. For binaries, I’m not so sure to be honest.

**Comment by hadronized** [comment]
I have decided to make binary releases (which I sign with my GPG key). I plan on making a specific documentation for packaging, so I will close this issue now, as I think packaging should leverage security concerns (guaranteeing users they are actually using something signed off by trustworthy people).

I will make a `PKGBUILD` for the [AUR](https://aur.archlinux.org) and will accept contributions to link to various release channels.

---

@rosingrind I have noticed you have not updated the `brew` formula in a while; please do not hesitate to ping me if you need help packaging / need a take over.

**Comment by hadronized** [comment]
Thank you @rosingrind. Just being curious there, what did you pick as new editing environment?

**Comment by rosingrind** [comment]
> what did you pick as new editing environment?

It's Zed and imo it's really good. Since some time ago I don't need terminal environments to develop anymore, so I gave it a try. It lacks for Linux/Windows support temporary though
